### ISSUE TITLE
Poor performance when dealing with multi-precision numbers

### ISSUE THREAD

**Issue by ChaosEternal** [cisco/ChezScheme#14]
* The test Program:

;; the fib iterative
(import (rnrs))

(define (fib n)
  (define (iter a b c)
    (cond ((= c 0) b)
          (#t (iter (+ a b) a (- c 1)))))
  (iter 1 0 n)
)

(display (fib 1000000))

* Run with Chez:

Chez Scheme Version 9.4
Copyright 1984-2016 Cisco Systems, Inc.

$ time scheme --program fib.scm  >/dev/null

real    0m23.007s
user    0m22.956s
sys     0m0.076s

* Run with vicare
Vicare Scheme version 0.4d0, 64-bit
Build 2015-10-19

Copyright (c) 2006-2010 Abdulaziz Ghuloum and contributors
Copyright (c) 2011-2015 Marco Maggi and contributors

$ time vicare fib.scm  >/dev/null                                                                                                                                             

real    0m6.372s
user    0m5.900s
sys     0m0.468s

**Comment by ChaosEternal** [comment]
lrwxrwxrwx 1 chaos chaos      6 Apr 27 10:59 petite -> scheme

**Comment by hyln9** [comment]
This is not strange.
Run same program under petite, you'll find that the running time is not changed so much.
Therefore it's not a compiler's issue but runtime library's.
Chez Scheme's bignum performance is indeed not as good as Vicare's which is powered by GMP.

**Comment by NalaGinrut** [comment]
@hyln9 what's the reason not use GMP? maybe because of licenses?

**Comment by ChaosEternal** [comment]
@hyln9 Thanks for point out it.

I updated the test program:

```
(import (rnrs (6)))


(define (fib n)
  (define (iter a b c)
    (cond ((= c 0) b)
          (#t (iter (mod (+ a b) 10232123) a (- c 1)))))
  (iter 1 0 n)
)

(display (fib 1000000000))
```
and the output becames:
```
$ time scheme --script fib.scm
8461706
real    0m15.524s
user    0m15.528s
sys     0m0.008s
```
```
$ time vicare --r6rs-script fib.scm                                                                                                                                           

real    0m24.490s
user    0m24.480s
sys     0m0.028s
```

**Comment by yinwang0** [comment]
You shouldn't be discussing these performance comparisons here. This is not even a code issue. This is distracting people from real issues. Not to mention that you haven't even done careful homework on how to properly do benchmark on compilers.

You are comparing Chez Scheme with Vicare, which is derived from Ikarus, also a very fast Scheme compiler. Every compiler is good at some optimizations but not others. Yes, maybe Chez Scheme is actually slower for this particular program, but can you say it has "poor performance"? No, the performance is definitely not poor. It may not be the best but it's not poor. And this doesn't need to be fixed. So this is not an issue.

If you do such comparisons with random code you wrote, and report these as issues to slower Scheme implementations such as MIT Scheme, or you compare with Python, their issue system will blow out real soon ;)

We should set some rules as what are legitimate issues.

**Comment by dybvig** [comment]
I'm actually happy to see issues related to performance.  Good performance is, after all, one of our primary goals.  It's true that a thorough comparison would involve a lot more and more careful benchmarking, but it's nice to have examples where we aren't doing as well as we could, especially if it motivates contributors to submit fixes.

Using gmp would indeed speed up programs that operate on large numbers, particularly programs that multiply and divide very large numbers.  Licensing is the primary issue blocking us from using it.  Another is that our own, different representation of bignums is baked into the compiler and run-time system in a couple of ways.

**Comment by NalaGinrut** [comment]
@dybvig Thanks for explaining! Now that Chez has its own bignum system, maybe it's unnecessary to do porting contribution for it, it's just need polishing if any possible. That's the reason why I ask.

**Comment by ChaosEternal** [comment]
The original motivation makes me submitting this issue is that I am suspecting whether the code released here is the same as  Chez used to be, blame me.

Thanks to @hyln9 's help, I finally realize that the performance difference is introduced by GMP. 
So i changed the subject to the current one. If it subject is still inappropriate, I can change it to anything else.

But I agree with @dybvig that the issue is valid and awaits to be resolved.

**Comment by ChaosEternal** [comment]
I think gmplib is LGPLv3 hence it is compatible with apache2.0 license.

**Comment by rain-1** [comment]
libgmp is an extremely large and complex body of code that would weigh down the rest of the system, making it larger, compiling would take longer. I don't think it is necessarily a good idea to include it. Are you planning on implementing this @ChaosEternal? If so I would urge you to consider other bignum libraries.

**Comment by dybvig** [comment]
Apache 2.0 licensed software can be used in LGPLv3 licensed projects, but not the other way around.

**Comment by johnwcowan** [comment]
For what it's worth, the Scheme48 library (also used in the Chicken numbers egg) is written in Scheme.

**Comment by yinwang0** [comment]
Sorry about my comment. I have seen many biased and premature benchmark comparisons recently, so I tend to be sensitive such things may get into the issue system, converting github into a discussion forum ;)

It seems this one is sorta valid question about bignums. I'd still suggest removing the word "poor" from the title, because the performance is not poor as compared to similar languages with bignum implementations (such as Java).


### Bignum results

Just for interest, I implemented the semantically same iterative fib function in three languages: Scheme, Java(using java's `BigInteger` class) and Python.

```racket
(define fib
  (lambda (n)
    (define iter
      (lambda (a b n)
        (cond
         [(= n 0) b]
         [else
          (iter (+ a b) a (- n 1))])))
    (iter 1 0 n)))
```

And the running results are quite interesting. I couldn't get Vicare built properly on the Mac after fixing several things in its build system, but it is essentially just Ikarus.

Here are the results running (fib 1000000) on my Macbook Pro.

```
Ikarus 0.0.3:   7.69s
Chez Scheme:   16.69s
Petite Chez:   17.66s
Java 8:        15.71s
Python:         9.86s
```

So Chez's bignum performance is similar to that of Java. It's interesting how Python stands out in performance for bignums. 

The the performance of bignums in Ikarus is really due to GMP. I took a look at the code of GMP and there is lots of processor specific optimizations done using assembly language and fancy instructions! Given how rarely people want super fast bignum arithmetic, I doubt Chez Scheme would want to have all that nasty details in the compiler ;)


### Recursive fib results

But if you write a recursive implementation of fib:

```racket
(define fib
  (lambda (n)
    (cond
     [(< n 2) n]
     [else
      (+ (fib (- n 1)) (fib (- n 2)))])))
```

and run (fib 48), you will get very different results.

```
Ikarus 0.0.3:   52.10s
Chez Scheme:    36.55s
Petite Chez:   409.53s
Java 8:         20.39s
Python:         >10 mins and still not finished
```

Take a look at Python's performance and compare with its bignum times :) 

Do you want super fast bignum arithmetic that normal programs rarely use, and sacrifice everyday performance on function calls, data structures, closures? After all, you can still call GMP and fast C code for such rare computations with FFI, so I don't think this is a big issue.

**Comment by hyln9** [comment]
@yinwang0 Agree. People often misunderstand where Chez's progressiveness lies, so toy code benchmark seems popular :)
However, efficient numeric tower operation is also essential. For bignum addition, GMP indeed wins on hardware specific optimization which is not suitable for.current Chez, but for multiplication and division, Chez's algorithm is already not desirable.

**Comment by yinwang0** [comment]
There might be a way...

I found that Ruby and Python both use GMP, which makes them quite fast in this test. But Ruby is BSD license, and they seem to be able to use call GMP, which is LGPL.

From this line, you can see that Ruby is using GMP:
>https://github.com/ruby/ruby/blob/trunk/doc/NEWS-2.1.0#L73

But from `doc/ChangeLog-1.9.3` (line 14778), you can see that they removed all LGPL code from the source:
> https://raw.githubusercontent.com/ruby/ruby/trunk/doc/ChangeLog-1.9.3

Since Chez Scheme is Apache license which is very similar to BSD, I wonder if Chez Scheme can do the same thing as Ruby, link to GMP library which the users install by themselves?

**Comment by dybvig** [comment]
That might be an option.  It would require conversions from Chez Scheme's representation to GMP's and back for each call into GMP and presumably used only in cases where that overhead is justified.  Or Chez Scheme's representation could be changed to match GMP's if GMP's representation is part of the public interface and the change doesn't add undue overhead of some other sort.

**Comment by lemaster** [comment]
Gambit Scheme's Bignums are pure scheme and quite fast (not as fast as GMP right now but it has been faster on certain operations for brief periods in the past). Maybe porting their algorithms would be better than adding a dependency on GMP.

**Comment by yinwang0** [comment]
I just tried the bignum fib program on Gambit, and it's slower than Chez. I'm not sure if I got the right way of running it, but running both `gsi fib.ss` and `gsc fib.ss; gsi fib.o1` got similar results. It takes at least twice as much time as Chez.

**Comment by ChaosEternal** [comment]
I have run my second program with chicken scheme, it cost around 1 minutes to finish as a compiled chicken program.

**Comment by lemaster** [comment]
You have to add some declarations to have gambit actually enable most of its optimizations. At minimum (declare (block) (standard-bindings)). It also depends on how you built gambit's own runtime.

A micro benchmark like this might be testing the allocator more than the bignum routines anyway.

**Comment by yinwang0** [comment]
I also wonder if I enabled Chez Scheme's optimizations properly. Any suggestions?

**Comment by lemaster** [comment]
My only real suggestion is, to properly test bignum implementations head to head, it makes sense to find some real bignum benchmarks. Like I said above, this fib function is properly more aaffected by the allocation and grabage collection policies than the implementation of bignum+

**Comment by soegaard** [comment]
Aubrey Jaffer's note "The Distribution of Integer Magnitudes in Polynomial Arithmetic" is worth a read:  http://people.csail.mit.edu/jaffer/CNS/DIMPA

Using a symbolic algebra system he examines the sizes of the integers used in the program. The conclusion is that it is the performance of small bignums that is the most important. He adds:

> The large reduction in frequency of occurrence versus bit-length means that small improvements in asymptotic running times for exotic bignum algorithms would bring negligible benefit in running this computer algebra program.

**Comment by NalaGinrut** [comment]
I have similar opinion, the fib example is more like a GC stress test. 
Guile (use libgmp for bignum) will be faster for fib if disabled GC. The note given by @soegaard reveals that Guile will do more collecting work if GC enabled, which seems the main reason of the loss.
```
If gc is activated however, guile performs the calculation within 200KB and calls gc several times. The whole thing takes about 25 seconds (!),
```
I don't know the difference of GC strategy between Chez and Vicare yet. But obviously this fib issue contains two part:
1. operation efficiency of bignum (related to bignum system)
2. alloc/collect efficiency of bignum (related to GC)

IMO, it is too early to come to a conclusion about the bottleneck before deeper research.

**Comment by NalaGinrut** [comment]
Well, I have some results after simple research.
I was planning to disable GC in Chez, but seems no option for it. Fortunately, Chez provides useful functions for tweaking GC.
The original result is:
```
    1749 collections
    17.650081080s elapsed cpu time, including 0.096985402s collecting
    17.651558134s elapsed real time, including 0.099052668s collecting
    43407533616 bytes allocated, including 43366316784 bytes reclaimed
```
The I tweaked these two:
```
(collect-trip-bytes (* 10 (collect-trip-bytes))) ; enlarge the allocate block each time
(collect-generation-radix 1000000) ; set to very large to avoid collect frequently
```
Then the result is:
```
    439 collections
    18.025043206s elapsed cpu time, including 0.081797687s collecting
    18.026762123s elapsed real time, including 0.082575259s collecting
    43406709264 bytes allocated, including 43316982624 bytes reclaimed
```

Well, the collection and the cost time reduce effectively, but the cpu time seems almost unchanged.
I don't know if it's the reasonably tweaking, but if it's correct for decreasing GC's activity, then my assumption is wrong: GC has little effect on this issue, and bignum operation matters.

**Comment by akeep** [comment]
@NalaGinrut You can disable the collector using the ``collection-request-handler`` (from the 8.4 version of CSUG, chapter 13---http://scheme.com/csug8/smgmt.html#./smgmt:s16):

> Automatic collection may be disabled by setting collect-request-handler to a procedure that does nothing, e.g.:
> ```
> (collect-request-handler void)
> ```

However, as the collection time is fairly small I would hazard to guess it is actually the bignum arithmetic that is killing this.  I suspect we'll have to improve the bignum operations in number.c to improve the results of this.

**Comment by yinwang0** [comment]
The fib test will indeed stress the GC a little, but this will only make the bignum calculation time worse for Chez Scheme, because Ikarus spends a lot more time in GC (1.5s Ikarus vs 0.1s Chez). Actually I doubt if you can do bignum benchmarks without using GC because bignums are allocated in the heap.

To be sure what we are talking about, this is the code:

```
(define fib-it
  (lambda (n)
    (define iter
      (lambda (a b n)
        (cond
         [(= n 0) b]
         [else
          (iter (+ a b) a (- n 1))])))
    (iter 1 0 n)))
```

It is iterative (tail recursive), so in principle we could allocate all stuff on stack and don't really need heap allocations or GC. But GC does happen because bignums are actually stored in the heap. I guess we can actually put them on the stack with the help of escape analysis.

Running (fib-it 1000000) cost Chez only 0.1 seconds in GC time (thanks to the generational collector?). So most of the time (17s) was spent in bignum calculations.

Running bignum test in Chez Scheme (0.1s GC time):

```
> (time (display (< 1 (fib-it 1000000))))
#t(time (display (< 1 ...)))
    1750 collections
    17.173180000s elapsed cpu time, including 0.100478000s collecting
    17.178098000s elapsed real time, including 0.102999000s collecting
    43407456912 bytes allocated, including 43407730416 bytes reclaimed
```

In comparison, Ikarus spends significantly more time (1.5s) in GC:

```
> (time (display (< 1 (fib-it 1000000))))
#trunning stats for (display (< 1 (fib-it 1000000))):
    10421 collections
    7171 ms elapsed cpu time, including 1512 ms collecting
    7172 ms elapsed real time, including 1521 ms collecting
    43402029744 bytes allocated
```

This means that Chez Scheme has a superior garbage collector than Ikarus. To be sure, you can test Ikarus with a simple factorial program, but written in two different ways (recursive and iterative):

Recursive factorial (caused 8.9s GC time):
```
(define fact
  (lambda (n)
    (cond
     [(= n 0) 1]
     [else (* n (fact (- n 1)))])))
```

```
> (time (< 1 (fact 200000)))
running stats for (< 1 (fact 200000)):
    9518 collections
    15006 ms elapsed cpu time, including 8915 ms collecting
    15013 ms elapsed real time, including 8930 ms collecting
    38616435792 bytes allocated
```

Iterative factorial (caused only 3s GC time):

```
(define fact-it
  (lambda (n)
    (define fact1
      (lambda (n prod)
        (cond
         [(= n 0) prod]
         [else (fact1 (- n 1) (* n prod))])))
    (fact1 n 1)))
```

```
> (time (< 1 (fact-it 200000)))
running stats for (< 1 (fact-it 200000)):
    10425 collections
    9874 ms elapsed cpu time, including 3020 ms collecting
    9883 ms elapsed real time, including 3035 ms collecting
    42222965144 bytes allocated
```

Running both fact and fact-it in Chez, you won't notice much difference in GC time. Both are about 0.1s!

```
> (time (< 1 (fact 200000)))
(time (< 1 ...))
    195 collections
    25.422836000s elapsed cpu time, including 0.110350000s collecting
    25.465419000s elapsed real time, including 0.110865000s collecting
    38622551408 bytes allocated, including 38222576752 bytes reclaimed
```

```
> (time (< 1 (fact-it 200000)))
(time (< 1 ...))
    390 collections
    27.385098000s elapsed cpu time, including 0.115547000s collecting
    27.442736000s elapsed real time, including 0.116543000s collecting
    42224402672 bytes allocated, including 42222535536 bytes reclaimed
```

So indeed Ikarus's garbage collector is not as good as Chez's, which makes bignum calculation time even worse for Chez ;) Since factorial uses multiplication, we can see GMP does make a big difference.

I agree that we need a better bignum benchmark, but a simple example seems good enough to demonstrate what's going on.

**Comment by hyln9** [comment]
@lemaster @NalaGinrut @akeep @yinwang0 
Firstly I think for multiplication and division there is no maze.
For addition I quickly wrote a benchmark for GMP:
bignum_add.c:
```
#include <stdio.h>
#include <gmp.h>

int main()
{
  mpz_t a, b, c;
  mpz_init(a);
  mpz_init(b);
  mpz_init(c);
  mpz_ui_pow_ui(a, 2, 1000000);
  mpz_ui_pow_ui(b, 2, 1000000);
  printf("start!\n");
  for(int i = 0; i < 1000000; i++)
  {
    mpz_add(c, a, b);
  }
}
```
We can hardly do in-place addition  in scheme, so allocation is more a problem than collection because we can combine it in GC. However we do not know allocation time in Chez Scheme, but seems it can be ignored accoding to @yinwang0 .
For clarification, scheme version is also available:
```
(run-cp0 (lambda (cp0 x) x))
(optimize-level 3)
(let [(a (expt 2 1000000))
      (b (expt 2 1000000))]
  (time
    (do [(i 0 (+ i 1))
         (c 0 (+ a b))]
        [(= i 1000000)])))
```
On my machine, results are as below:
GMP: (time before "start" is too short so included)
```
real	0m10.100s
user	0m10.096s
sys	0m0.000s
```
Chez:
```
    2000 collections
    55.476879137s elapsed cpu time, including 0.244292071s collecting
    55.481766220s elapsed real time, including 0.247766474s collecting
    125025280000 bytes allocated, including 125028249216 bytes reclaimed
```
So addition performance is still low.

**Comment by yinwang0** [comment]
I wrote a similar Java program doing `(fact 200000)`. It shows that Java's BigInteger has better multiplication performance than Chez's. It took Java 12.8s. 

In summary:
```
Ikarus: 9.9s
Java: 12.8s
Chez: 27.4s
```

GMP is still the best among the three, for obvious reasons.

**Comment by hyln9** [comment]
Another interesting thing I observed is that if we use higher `optimize-level`, Ikarus can perform DCE on my benchmark code above while Chez cannot whether in REPL or command line, with "(time ..." or without, with "--program" or without.

**Comment by yinwang0** [comment]
What does "DCE" stand for?

**Comment by hyln9** [comment]
@yinwang0 "dead code elimination", my bad.

**Comment by akeep** [comment]
As we think about trying to improve the bignum implementation in Chez, it might be interesting to take a look at the Glasgow Haskell Compiler (GHC) work around the same issue: https://ghc.haskell.org/trac/ghc/wiki/ReplacingGMPNotes.

They started from a different point, in that GHC had been using the GNU MP, but there were uses of GHC for whom the LGPL licensing of GNU MP was problematic (see: https://ghc.haskell.org/trac/ghc/ticket/601).  They experimented with a number of other fast arbitrary precision libraries with more permissive licenses, as well as creating a "fast enough" implementation in Haskell.  One (or more) of these more permissively licensed libraries might be an interesting option for improving Chez's bignum performance.  As of 7 years ago (when the GHC community went through this exercise), there were quite a few interesting trade offs in performance given the libraries available at the time.  I found an additional library, [bigz](https://sourceforge.net/projects/bigz/), listed on the [wikipedia arbitrary precision page](https://en.wikipedia.org/wiki/List_of_arbitrary-precision_arithmetic_software).  Since it has been 7 years since the GHC community did these experiments, it is probably worth evaluating the options again.

@hyln9 I'll have to look at the DCE as a separate issue.

**Comment by hyln9** [comment]
@akeep Nice information. Personally, I don't like GNU MP for its license. If we want to make Chez reliable and clean, I'd prefer directly improving implementations in `number.c` instead of being dependent on 3rd party libraries. I can help with it if appropriate.

**Comment by yinwang0** [comment]
In the end, since bignums are used so rarely in everyday programming, I think it's enough a programming language's numbers correspond to whatever the machine has (int32, int64, float, double...). Users who need bignums should handle it themselves or call a library of their choice.

Building the bignums seamlessly into a programming language seems hardly satisfy user's needs, and it introduces unnecessary overhead. Once the users need bignums, they are going to be picky about whatever the language provides. Eventually they may end up calling a third-party library such as GMP which they found faster than the built-in implementation.

The original idea of the numeric tower is to ease the users from thinking about difference of numbers, but it seems once they need such things, they better do some extra work :)

**Comment by johnwcowan** [comment]
@yinwang0: This is Scheme, where fixnum overflow (using the term "fixnum" broadly) can return an exact non-fixnum, or an inexact approximation, or raise an implementation-restriction error, depending on the implementation.  This is true of all standards from R4RS onwards at least, and most .  See http://trac.sacrideo.us/wg/wiki/NumericTower for what many existing Schemes actually do.

**Comment by yinwang0** [comment]
@johnwcowan That's just a general comment across all languages and how often I encounter bignums over the years. I understand that it must be there in Scheme, just maybe not a priority of making it efficient. But well, I'm happy somebody already volunteered to try things out :)

**Comment by NalaGinrut** [comment]
@hyln9  @akeep I think this fib case is too simple to take advantage of DCE, so maybe raise another issue?

**Comment by NalaGinrut** [comment]
Personally, I'm not interested in porting libgmp to Chez. Because one of limits of libgmp is that very big number operation is optimized to be faster while small number is slower, which is a compromise obviously. 
For Chez, I think the strategy could be less extreme. Maybe we could find a way to improve big number operation in a limited expectation (say, not as fast as libgmp), but it shouldn't drag the small number operation.
Well, I confess it sounds idealism. Anyway, if we can't find a way to improve big number without effecting small number, I prefer keep it as it is.

**Comment by hyln9** [comment]
@NalaGinrut 

Chez obviously do DCE, so it is whether a bug or misunderstanding, rather than a performance issue which needs complex benchmark.

As far as I know, there is no evidence which shows that GMP has drawbacks on relatively small (i.e. mid-size) numbers. I agree with your opinions on performance but fixnum is managed by compiler itself without overhead while there are numbers of algorithms or strategies on mid-size bignum operations.

**Comment by NalaGinrut** [comment]
@hyln9 
Alright, I've taken a look again, GMP use different algorithm for different operand size. So maybe not the case I'm afraid of.

**Comment by yinwang0** [comment]
I looked at the code of Chez's number.c and wonder if there is a simple way of make addition faster for 64-bit machines.

I see that the `bigit` type is always defined as U32, and `big_add` operates on bigits, so it's not going to use 64-bit addition instruction, thus is doing twice as many additions.

**Comment by hyln9** [comment]
@yinwang0 In addition, since adc instruction is widely available, the EADDC macro is unnecessary. In the end we are further and further from standard c…

**Comment by yinwang0** [comment]
@hyln9 Indeed, it is the ADC instruction doing the trick in GMP. I played with GMP's configuration again, and found that Ikarus can't use GMP's 64-bit ABI. I had to compile GMP with ABI=32, so Ikarus was not utilizing the 64-bit instructions. But it's still a lot faster because it's using x86's ADC instruction.

The code that is in actual use by Ikarus on my machine is in this file:

```
mpn/add_n.asm -> ../mpn/x86/p6/aors_n.asm
```

The assembly code looks like this, where ADCSBB is defined to be adc earlier. It's just repeatedly calling adc.

```
L(top):
	jecxz	L(end)
L(ent):
Zdisp(	mov,	0,(up,n,4), %eax)
Zdisp(	ADCSBB,	0,(vp,n,4), %eax)
Zdisp(	mov,	%eax, 0,(rp,n,4))

	mov	4(up,n,4), %edx
	ADCSBB	4(vp,n,4), %edx
	mov	%edx, 4(rp,n,4)

	mov	8(up,n,4), %eax
	ADCSBB	8(vp,n,4), %eax
	mov	%eax, 8(rp,n,4)

	mov	12(up,n,4), %edx
	ADCSBB	12(vp,n,4), %edx
	mov	%edx, 12(rp,n,4)

	mov	16(up,n,4), %eax
	ADCSBB	16(vp,n,4), %eax
	mov	%eax, 16(rp,n,4)

	mov	20(up,n,4), %edx
	ADCSBB	20(vp,n,4), %edx
	mov	%edx, 20(rp,n,4)

	mov	24(up,n,4), %eax
	ADCSBB	24(vp,n,4), %eax
	mov	%eax, 24(rp,n,4)

	mov	28(up,n,4), %edx
	ADCSBB	28(vp,n,4), %edx
	mov	%edx, 28(rp,n,4)

	lea	8(n), n
	jmp	L(top)
```

**Comment by hyln9** [comment]
@yinwang0 
GMP's implementation of addition above combines adc with loop unrolling, which increases the ratio of arithmetic instructions. Since GMP's configuration script probably failed to detect your cpu type correctly, more optimization method such as SIMD can be expected. On the other hand 64bit and 32 bit adc have same throughput on x86_64, therefore double the potential performance.

**Comment by dybvig** [comment]
It would be great for someone to add the requisite ifdefs and asm instructions to use more efficient operators, and for someone to implement fancier algorithms, but please be careful in the process not to copy (or even study) code from GNU-licensed systems like GMP and Ikarus.  Though Aziz would probably be willing to put the portions of the code you want to use under the Apache 2.0 license.

**Comment by hyln9** [comment]
@dybvig 
Yeah, that the reason why I haven't look at GMP's code yet except for the above one, so I cannot know GMP's algorithms. My major references is wikipedia and academic papers (since I am an undergraduate I have access).
We need a clean room implementation.

**Comment by yinwang0** [comment]
@hyln9 GMP detected my processor as Haswell correctly and configured to use 64bit instructions and registers. It's just Ikarus 0.0.3 that can't use 64bit. So it's using adc with unrolling only with addition. It's using some mmx and sse instructions with other operations. Addition seems fast enough with just ADC and easy to implement with embedded asm.

**Comment by hyln9** [comment]
@yinwang0 It seems that there are no SIMD instructions for adding with carry for x86 yet.

**Comment by yinwang0** [comment]
I guess there can't be SIMD addition with carry because the carry propagation is hard to do in parallel with bignums.

**Comment by yinwang0** [comment]
http://stackoverflow.com/questions/27923192/practical-bignum-avx-sse-possible

**Comment by NalaGinrut** [comment]
It sounds like we have to rewrite all bignum operations and maintain related assembly code for all supported platforms, which is  alreadydone by libgmp. 
Take an example of addition operation. If we don't care about x86, maybe just rewrite EADDC is enough (no?). For X86 or certain platform doesn't contain necessary AVX in the future, it's necessary to keep current EADDC.
Well, maybe libgmp is still an option.

**Comment by yinwang0** [comment]
I wrote a small demo program using GCC's extended assembly and the ADC instruction.

  https://gist.github.com/yinwang0/290f34bb567a896eada4745173aa4477

The main part of the demo is exactly following the names in `big_add_pos` so it's easy to swap the code in Chez Scheme.

The demo itself seems to be correct, but after swapping the code into `big_add_pos` and rebuild Chez Scheme after "make clean", I got some error saying "nonrecoverable invalid memory reference".

It looks like a good starting point. I'm going to look more, but to make the development "parallel", some of you may want to try and may finding out how to make it work with Chez Scheme.

**Comment by yinwang0** [comment]
@NalaGinrut It looks Scheme doesn't provide that many bignum operators, and the open sourced Chez Scheme doesn't support that many architectures. I think it's worthwhile enough if we can just make the few operations fast under x86, because that's what most people use.

**Comment by johnwcowan** [comment]
The operations on bignums that are needed are those listed in section 11.7.4 of R6RS that are meaningful when applied to exact integers, namely `inexact = < > <= >= zero? positive? negative? odd? even? max min + - * / div mod div-and-mod div0 mod0 div0-and-mod0 gcd lcm expt number->string string->number`.

**Comment by NalaGinrut** [comment]
@yinwang0 I haven't read the code carefully, but the inlined assembly code contains side-effects, it is  generally to add ```volatile``` or ```__volatile__``` to avoid the compiler breaking assembly code block while optimizing.
https://www.ibiblio.org/gferg/ldp/GCC-Inline-Assembly-HOWTO.html#ss5.4
Or it may not safe:
```
[nalaginrut@debian:tmp] cc ba.c -o ba
[nalaginrut@debian:tmp] ./ba
z[0] = 1
z[1] = 0
z[2] = 0
z[3] = 0
z[4] = 0
[nalaginrut@debian:tmp] cc ba.c -o ba -O3
[nalaginrut@debian:tmp] ./ba
z[0] = 0
z[1] = 0
z[2] = 0
z[3] = 0
z[4] = 0
```

@johnwcowan Maybe we don't have to optimize them all, and it is necessary to do more benchmarks first.

**Comment by yinwang0** [comment]
@NalaGinrut Indeed not having `__volatile__` is one of the reasons that it's not right. Adding `__volatile__` and I see other kinds of issue. But it seems I'm getting closer to get it work.

**Comment by alex-shpilkin** [comment]
@NalaGinrut @yinwang0 On an inline assembly block, `volatile` basically means that the block has effects other than those explicitly indicated to GCC using the extended asm syntax input and output specifiers; indeed one of the reasons for the extended syntax is to allow GCC to do CSE, DCE and similar optimizations on inline assembly blocks. Writing `asm volatile` is basically giving up on explaining to the compiler what the assembly block does (think `mov cr3`); I don’t see how this should be required here. A smaller but still probably excessive hammer is the `memory` output.

On the other hand, it looks like the function signature may need some `restrict` specifiers (i.e. the code doesn’t work correctly unless the pointers don’t alias each other), but I’ll leave that to the person that wrote it. (If you require GCC/ICC/clang you might as well assume support for `restrict` I think.)

**Comment by yinwang0** [comment]
@alex-shpilkin I don't see why it's needed there but I got wrong results with gcc optimizations (-O1 and above) if there is no `__volatile__`. Maybe that's a hint why the code is still not correct.

**Comment by yinwang0** [comment]
I fixed bugs in my code and now it can be built with Chez and produce correct results. The `__volatile__` can be removed and still produce correct results with `-O3`.

The code is here in my fork:

> https://github.com/yinwang0/ChezScheme/commits/improve-big-add

It's basically a translation of Chez's original code, but using ADC instruction to handle the carry.

Unfortunately the performance of the assembly code is not even as good as Chez's original C code. The timing is about 28s (my change) vs 17s (chez's original). So I'm no longer sure about ADC instruction is the real reason that GMP is fast. We could do some loop unrolling and see if that helps.

**Comment by yinwang0** [comment]
Slowness fixed. That's because I used x86's `loop` instruction. Changed loop control to use `jecxz` and `lea`, and now it's faster than Chez's original code, although not much: 14s vs 18s.

I have no idea why `loop` instruction is so slow, maybe just old circuit leftover in the processor ;)

**Comment by alex-shpilkin** [comment]
@yinwang0 Indeed, as [Agner Fog’s tables](http://www.agner.org/optimize/instruction_tables.pdf) show,  `LOOP` (which existed in 8086 already) is a whopping 11 cycles on Sandy Bridge compared to two cycles for `JECXZ` (introduced in 80486 IIRC). In turn, `JECXZ` is not needed, because (recent Intel) processors are able to fuse the `CMP`/`JNE` pair into one microinstruction internally.

Most probably the loop would benefit from unrolling, but the degree of it is to be determined empirically, preferably on benchmarks with realistic number magnitudes. (Would the tail dominate? Can we just declare that bignums are allocated in 4x32/64 chunks? Should we align to 128-bit memory fetch lines [C11 `alignas`]? &c.)

**Comment by yinwang0** [comment]
@alex-shpilkin I unrolled the loops and adjusted the way for indexing into the arrays. Now the timing is reduced to about 10.5s (as compared to 18s).

I can't use `CMP / JNE` because `CMP` may change the `CF` flag, which `ADC` will need. It's quite some tricks that it does by using `LEA` for addition into `ECX`, and `JECXZ` for looping.

The unrolling is done only 2x. I tried to unroll the loops using 4x32 chunks but it didn't improve over 2x32, so I'm currently keeping it at 2x just for clarity of the code. Some other things can be done for example using 64-bit addition. Also it needs some `#if`'s for targeting correct architecture.

**Comment by yinwang0** [comment]
@dybvig I got some endianness problems when trying to use 64-bit instructions. Since the 32-bit chunks of bignums are stored in a big-endian way, I got wrong byte order when using `movq` instruction of x86 which assumes little-endian.

I wonder if Chez already conditionally handles the endianness regarding bignums? Is there an easy variable to change so can I switch it to 64-bit mode?

**Comment by hyln9** [comment]
@NalaGinrut GMP has nothing to do with AVX.
@yinwang0 Currently bigit_bits is always 32 regardless of machine type.

**Comment by hyln9** [comment]
I'd like to discuss some design decisions here before going any further.

I'd prefer to use assembly source file directly instead of inline assembly which is not portable (e.g. x86_64 inline assembly is not available in msvc) and stable (depends on compiler behavior). We can create a folder called "a" and put assembly code into different folders distinguished by machine types.

In c sources, we can use ifdefs to tell whether a function is implemented in asm or c.

**Comment by yinwang0** [comment]
@hyln9 Indeed inline assembly is not as good as separate assembly files. My implementation can be considered a proof of concept that ADC can run faster. GCC's inline assembly is also quirky to get right, and is thus just for small snippets. It's better to use a portable assembly language line NASM to write those.

**Comment by NalaGinrut** [comment]
@hyln9 
I thought we may choose AVX for bignum, but seems it's more inconvenient than ADC.
And libgmp proves ADC is enough. My bad English.
I agree with you to create folder for assembly code. But maybe we could find better way to optimize the current prove of concept first. It is not as fast as expected.

**Comment by yinwang0** [comment]
@NalaGinrut You are right. My code is still not as fast as GMP's although they look very similar, so you may want to play with it and see if there is easy way to improve. I also unrolled it 4 and 8 times, but it didn't improve the performance. I don't know yet what other tricks are in GMP for addition.

**Comment by hyln9** [comment]
@yinwang0 After a quick look at `numeric.c`, I found that maybe the copy operation of `copy_normalize` is redundant at least for bignum addition. I cannot look into it before my boring homework is finished, though.

**Comment by yinwang0** [comment]
Tried to use ROL and ROR for switching endians, but unfortunately they change the carry flag CF. Keep in mind the trick with ADC lies in that you can't use instructions that can change the CF flag between the ADCs.

So it seems we really need to store the bignum in different order with 64-bit machines.

**Comment by hyln9** [comment]
@akeep After some investigations into Ikarus, I think the DCE issue is actually a loop-invariant hoisting one.

**Comment by yinwang0** [comment]
Actually I don't see where this code is related to DCE.

**Comment by hyln9** [comment]
@yinwang0 It has no return value.

**Comment by yinwang0** [comment]
@hyln9 Are we talking about the same piece of code?

```
(define fib-it
  (lambda (n)
    (define iter
      (lambda (a b n)
        (cond
         [(= n 0) b]
         [else
          (iter (+ a b) a (- n 1))])))
    (iter 1 0 n)))

(time (display (< 1 (fib-it 1000000))))
```

I don't see where the dead code is.

**Comment by hyln9** [comment]
@yinwang0 Sorry for the confusion I made.
I meant this: https://github.com/cisco/ChezScheme/issues/14#issuecomment-215851686

**Comment by dybvig** [comment]
I'm going to take the opportunity in the lull between posts to close this issue and suggest that further discussion either be held privately amongst the thread contributors or on the mailing list.
